[BondedFinance] Add benchmarks

c410-f3r opened the pull request:
Pallet benchmarks are intended to be used as an upper-bound for extrinsic calls, which means creating benchmark functions with worst case scenarios.

Weight is a unit of time and every signed extrinsic should have a weight to calculate fees or the average number of transactions for a block. The same necessity is true for unsigned transactions because theoretically anyone could submit infinite number of extrinsics inside a single block for anything that has `#[weight = 0]`, therefore, please consider adding weights for non-root calls of the `ping` pallet to avoid third-party flood.

c410-f3r commented on frame/bonded-finance/src/lib.rs:
```diff
@@ -115,10 +117,11 @@ pub mod pallet {
 		/// The native currency, used for the stake required to create an offer.
-		type NativeCurrency: FungibleTransfer<AccountIdOf<Self>>;
+		type NativeCurrency: fungible::Mutate<AccountIdOf<Self>>
+			+ fungible::Transfer<AccountIdOf<Self>>;
 
 		/// The multi currency system offers are based on.
-		type Currency: FungiblesTransfer<AccountIdOf<Self>>;
+		type Currency: fungibles::Mutate<AccountIdOf<Self>> + FungiblesTransfer<AccountIdOf<Self>>;
```
There is NO way to mint generic tokens without these bounds. IMO, it is not a problem because the runtime implementation already implements them

hussein-aitlahcen replied:
AFAIK the pallet is not minting any token as both the bond asset and the reward are provided by users interacting with it. The reward is provided by the bond offer issuer and the bond asset is provided by users willing to bond?

c410-f3r replied:
As well as unit tests, benchmarks also need initial minting in order to test internal logic. The different between both is that unit tests interacts with concrete types while benchmarks have to figure out abstract programming in order to measure different `runtime` performances, thus the reason of `fungibles::Mutate` along side the already existing `fungibles::Inspect` (alias of `FungiblesTransfer`).

hussein-aitlahcen replied:
Alright then, I thought it was about the pallet features. We could probably have the tokens in the genesis block using the ExtBuilder but it's not a big deal, thanks for clarifying :).

c410-f3r commented on frame/bonded-finance/src/lib.rs:
```diff
@@ -376,6 +375,10 @@ pub mod pallet {
 				}
 			})
 		}
+
+		pub(crate) fn account_id(offer_id: T::BondOfferId) -> AccountIdOf<T> {
+			T::PalletId::get().into_sub_account(offer_id)
+		}
```
This function is only used inside the crate for "pallet" and "benchmark" purposes. IMO, it does not make sense the leak pallet methods visibility for public usage

hussein-aitlahcen replied:
Good catch, was made public for tests of course.

c410-f3r commented on frame/bonded-finance/src/utils.rs:
```diff
@@ -0,0 +1,3 @@
+#[cfg(any(feature = "runtime-benchmarks", test))]
+
+pub const MIN_VESTED_TRANSFER: u32 = 1_000_000;
```
Created a new module that shares common information for "pallet" and "benchmark" usage. Although minimal, it also serves as a place for future additions.

c410-f3r commented on runtime/picasso/src/lib.rs:
```diff
@@ -773,6 +773,43 @@ impl assets::Config for Runtime {
 }
 
+#[cfg(feature = "develop")]
+parameter_types! {
+	pub const MaxVestingSchedule: u32 = 2;
+	pub const MinVestedTransfer: u64 = 1_000_000;
+}
```
Feel free to change these parameters

c410-f3r commented on runtime/picasso/src/lib.rs:
```diff
@@ -773,6 +773,43 @@ impl assets::Config for Runtime {
+}
+
+#[cfg(feature = "develop")]
+parameter_types! {
+	pub const BondedFinanceId: PalletId = PalletId(*b"bondedfi");
+	pub const MinReward: Balance = 1_000_000;
+	pub const Stake: Balance = 10_000;
+}
```
Feel free to change these parameters

c410-f3r commented on runtime/picasso/src/lib.rs:
```diff
@@ -773,6 +773,43 @@ impl assets::Config for Runtime {
+}
+
+#[cfg(feature = "develop")]
+impl vesting::Config for Runtime {
+	type Currency = Tokens;
+	type Event = Event;
+	type MaxVestingSchedules = MaxVestingSchedule;
+	type MinVestedTransfer = MinVestedTransfer;
+	type VestedTransferOrigin = system::EnsureSigned<AccountId>;
+	type WeightInfo = ();
+}
```
The `bonded-finance` pallet needs the `vesting` pallet, thus its runtime addition

c410-f3r commented on runtime/primitives/src/currency.rs:
```diff
@@ -9,7 +9,9 @@ use sp_runtime::{ArithmeticError, DispatchError, RuntimeDebug};
 use sp_runtime::sp_std::ops::Deref;
 
-#[derive(Encode, Decode, Eq, PartialEq, Copy, Clone, RuntimeDebug, PartialOrd, Ord, TypeInfo)]
+#[derive(
+	Encode, Decode, Eq, PartialEq, Copy, Clone, RuntimeDebug, PartialOrd, Ord, TypeInfo, CompactAs,
+)]
```
`CompacAs` is the `#[transparent]` counter-part of `parity-scale-codec` and is being used to decode assets for the `bonded-finance` benchmarks

hussein-aitlahcen replied:
TIL that, thanks :)

hussein-aitlahcen commented:
Any chance we have this PR rebased, I think the CI has been fixed in one of the latest commits :)

c410-f3r commented:
> Any chance we have this PR rebased, I think the CI has been fixed in one of the latest commits :)

Rebased

c410-f3r commented:
Benchmarks were added, code is compiling and tests are green. The only thing remaining is figure out why the runtime is panicking when issuing `cargo run  .... --features runtime-benchmarks ... -- benchmarks ...`

```
Error: Input("Error executing and verifying runtime benchmark: Other(\"Wasm execution trapped: wasm trap: unreachable\\nwasm backtrace:\\n    0: 0x37a4 - <unknown>!rust_begin_unwind\\n    1: 0x2bbd - <unknown>!core::panicking::panic_fmt::h8f11323637b4db3b\\n    2: 0x286d - <unknown>!core::result::unwrap_failed::h5b559a304463e509\\n    3: 0x1ca0aa - <unknown>!<picasso_runtime::Runtime as frame_benchmarking::utils::runtime_decl_for_Benchmark::Benchmark<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,u32>,picasso_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_spec_version::CheckSpecVersion<picasso_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<picasso_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<picasso_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<picasso_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<picasso_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<picasso_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<picasso_runtime::Runtime>)>>>>::dispatch_benchmark::hcaa3386e858e74d0\\n    4: 0x38b893 - <unknown>!Benchmark_dispatch_benchmark\\n\")")
2021-12-14 15:32:29.128 DEBUG main wasm-heap: allocator being destroyed, max_total_size 2568, max_bumper 1238872    
2021-12-14 15:32:29.128 TRACE main state: Return ext_id=67ea was_native=false result=Err(Other("Wasm execution trapped: wasm trap: unreachable\nwasm backtrace:\n    0: 0x37a4 - <unknown>!rust_begin_unwind\n    1: 0x2bbd - <unknown>!core::panicking::panic_fmt::h8f11323637b4db3b\n    2: 0x286d - <unknown>!core::result::unwrap_failed::h5b559a304463e509\n    3: 0x1ca0aa - <unknown>!<picasso_runtime::Runtime as frame_benchmarking::utils::runtime_decl_for_Benchmark::Benchmark<sp_runtime::generic::block::Block<sp_runtime::generic::header::Header<u32,sp_runtime::traits::BlakeTwo256>,sp_runtime::generic::unchecked_extrinsic::UncheckedExtrinsic<sp_runtime::multiaddress::MultiAddress<<<sp_runtime::MultiSignature as sp_runtime::traits::Verify>::Signer as sp_runtime::traits::IdentifyAccount>::AccountId,u32>,picasso_runtime::Call,sp_runtime::MultiSignature,(frame_system::extensions::check_spec_version::CheckSpecVersion<picasso_runtime::Runtime>,frame_system::extensions::check_tx_version::CheckTxVersion<picasso_runtime::Runtime>,frame_system::extensions::check_genesis::CheckGenesis<picasso_runtime::Runtime>,frame_system::extensions::check_mortality::CheckMortality<picasso_runtime::Runtime>,frame_system::extensions::check_nonce::CheckNonce<picasso_runtime::Runtime>,frame_system::extensions::check_weight::CheckWeight<picasso_runtime::Runtime>,pallet_transaction_payment::ChargeTransactionPayment<picasso_runtime::Runtime>)>>>>::dispatch_benchmark::hcaa3386e858e74d0\n    4: 0x38b893 - <unknown>!Benchmark_dispatch_benchmark\n"))
```

As you can see, the backtrace is not very useful and I already tried many things for several hours without success: Changed CLI parameters, changed benchmark logic, changed call inputs, changed runtime parameters, added logs, etc, etc, etc...

c410-f3r commented:
```
unused dependencies:
`pallet-bonded-finance v0.0.1 (/home/runner/_work/composable/composable/frame/bonded-finance)`
└─── dependencies
     └─── "frame-benchmarking"
Note: They might be false-positive.
      For example, `cargo-udeps` cannot detect usage of crates that are only used in doc-tests.
      To ignore some dependencies, write `package.metadata.cargo-udeps.ignore` in Cargo.toml.
```

`frame-benchmarking` is being used inside the `benchmarks` module. The warning was trigged probably because `frame-benchmarking/std` unconditionally (cargo stuff) brings the whole crate, i.e., a false-positive.

c410-f3r commented:
Not sure why CI is still failing (🤷) but I will take a look tomorrow

c410-f3r commented:
Should be good now

hussein-aitlahcen commented:
LGTM, @c410-f3r any chance we rebase this :) so we merge it?

c410-f3r commented:
> LGTM, @c410-f3r any chance we rebase this :) so we merge it?

Done

hussein-aitlahcen reviewed (approved):
Thanks a lot :)

c410-f3r commented:
Merry Christmas
